File: creational/typescript/prototype_structured/registry.ts
import { Prototype } from './interfaces';

export class PrototypeRegistry {
  private prototypes: Map<string, Prototype> = new Map();
  
  registerPrototype(key: string, prototype: Prototype): void {
    this.prototypes.set(key, prototype);
  }
  
  clone(key: string): Prototype | null {
    const prototype = this.prototypes.get(key);
    return prototype ? prototype.clone() : null;
  }
}

File: creational/dart/prototype_structured/registry.dart
import 'interfaces.dart';

class PrototypeRegistry {
  final Map<String, Prototype> _prototypes = {};
  
  void registerPrototype(String key, Prototype prototype) {
    _prototypes[key] = prototype;
  }
  
  Prototype? clone(String key) {
    var prototype = _prototypes[key];
    return prototype?.clone();
  }
}
